tsdb/store.go

```diff
@@ -672,6 +673,26 @@ func (s *Store) Shard(id uint64) *Shard {
 	return sh
 }
 
+// ClearBadShardList will remove all shards from the badShards cache
+// this will allow for lazy loading of bad shards if/when they are no
+// longer in a "bad" state. This method will return any shards that
+// were removed from the cache.
+func (s *Store) ClearBadShardList() map[uint64]error {
+	if s.badShards.shardErrors == nil {
+		badShards := make(map[uint64]error)
```

Comment:
I believe that this needs to be protected by the `s.badShards.mu` throughout this method.

`s.badShards.shardErrors` should never be `nil`, so we probably want to do two things:

1. Log a warning
2. Set `s.badShards.ShardErrors`  to an empty map.

Then we can proceed with the clone and clear code below which will be the same in all cases.

Reply:
`ClearBadShardList` should use `GetBadShardList` so that all the mutex and clone logic is localized to a single method.

Reply:
That would produce: 

```go
func (s *Store) ClearBadShardList() map[uint64]error {
	badShards := s.GetBadShardList()
	clear(s.badShards.shardErrors)

	return badShards
}

// GetBadShardList is exposed as a method for test purposes
func (s *Store) GetBadShardList() map[uint64]error {
	s.badShards.mu.Lock()
	defer s.badShards.mu.Unlock()

	if s.badShards.shardErrors == nil {
		s.Logger.Warn("badShards was nil")
		s.badShards.shardErrors = make(map[uint64]error)
	}

	shardList := maps.Clone(s.badShards.shardErrors)

	return shardList
}
```

So it's likely I would still need to use a mutex in `ClearBadShardList()` to call the `clear()` method on the original map?

Reply:
Yes, you must use the mutex around the `clear`